rpcserver.go

```diff
@@ -3509,57 +3509,109 @@ func (r *rpcServer) marshallRoute(route *routing.Route) *lnrpc.Route {
+			// a copy of the channel graph.
+			PubKey: hex.EncodeToString(hop.PubKeyBytes[:]),
 		}
 		incomingAmt = hop.AmtToForward
 	}
 
 	return resp
 }
 
+// unmarshallHopByChannelLookup unmarshalls an rpc hop for which the pub key is
+// not known. This function will query the channel graph with channel id to
+// retrieve both endpoints and determine the hop pubkey using the previous hop
+// pubkey. If the channel is unknown, an error is returned.
+func unmarshallHopByChannelLookup(graph *channeldb.ChannelGraph, hop *lnrpc.Hop,
+	prevPubKeyBytes []byte) (*routing.Hop, error) {
+
+	// Discard edge policies, because they may be nil.
+	edgeInfo, _, _, err := graph.FetchChannelEdgesByID(hop.ChanId)
+	if err != nil {
+		return nil, fmt.Errorf("unable to fetch channel edges by "+
+			"channel ID %d: %v", hop.ChanId, err)
+	}
+
+	var pubKeyBytes [33]byte
+	switch {
+	case bytes.Equal(prevPubKeyBytes, edgeInfo.NodeKey1Bytes[:]):
+		pubKeyBytes = edgeInfo.NodeKey2Bytes
+	case bytes.Equal(prevPubKeyBytes, edgeInfo.NodeKey2Bytes[:]):
+		pubKeyBytes = edgeInfo.NodeKey1Bytes
+	default:
+		return nil, fmt.Errorf("channel edge does not match expected node")
+	}
+
+	return &routing.Hop{
+		OutgoingTimeLock: hop.Expiry,
+		AmtToForward:     lnwire.MilliSatoshi(hop.AmtToForwardMsat),
+		PubKeyBytes:      pubKeyBytes,
+		ChannelID:        edgeInfo.ChannelID,
+	}, nil
+}
+
+// unmarshallHop unmarshalls an rpc hop that contains the hop pubkey. The
+// channel graph doesn't need to be queried because all information required for
+// sending the payment is present.
+func unmarshallHop(hop *lnrpc.Hop) (*routing.Hop, error) {
+	pubKey, err := hex.DecodeString(hop.PubKey)
+	if err != nil {
+		return nil, fmt.Errorf("cannot decode pubkey %s", hop.PubKey)
+	}
+
+	var pubKeyBytes [33]byte
+	copy(pubKeyBytes[:], pubKey)
+
+	return &routing.Hop{
+		OutgoingTimeLock: hop.Expiry,
+		AmtToForward:     lnwire.MilliSatoshi(hop.AmtToForwardMsat),
+		PubKeyBytes:      pubKeyBytes,
+		ChannelID:        hop.ChanId,
+	}, nil
+}
+
+// unmarshallRoute unmarshalls an rpc route. For hops that don't specify a
+// pubkey, the channel graph is queried.
 func (r *rpcServer) unmarshallRoute(rpcroute *lnrpc.Route,
 	graph *channeldb.ChannelGraph) (*routing.Route, error) {
 
-	node, err := graph.SourceNode()
+	sourceNode, err := graph.SourceNode()
 	if err != nil {
 		return nil, fmt.Errorf("unable to fetch source node from graph "+
 			"while unmarshaling route. %v", err)
 	}
 
-	nodePubKeyBytes := node.PubKeyBytes[:]
+	prevNodePubKey := sourceNode.PubKeyBytes[:]
 
 	hops := make([]*routing.Hop, len(rpcroute.Hops))
 	for i, hop := range rpcroute.Hops {
-		// Discard edge policies, because they may be nil.
-		edgeInfo, _, _, err := graph.FetchChannelEdgesByID(hop.ChanId)
-		if err != nil {
-			return nil, fmt.Errorf("unable to fetch channel edges by "+
-				"channel ID for hop (%d): %v", i, err)
+		var routeHop *routing.Hop
+
+		if hop.PubKey == "" {
```

Comment:
I think I would prefer this check be done inside the `unmarshal method`. Shouldn't be a problem passing in the graph even though it might not be needed.